content/client-lib-development-guide/features.textile

```diff
@@ -1059,6 +1063,19 @@ h3(#push-channels). Push channels
 
+h3(#local-device). LocalDevice
+
+* @(RSH8)@ In platforms that support receiving push notifications, the @device()@ method on the @AblyRest@ interface returns an instance of @LocalDevice@ that represents the current state of the device in respect of it being a target for push notifications.
+** @(RSH8a)@ The @LocalDevice@ is initialised when first required, either as a result of a call to @AblyRest.device()@, or as a result of an operation involving the Activation State Machine. The @LocalDevice@ @id@, @clientId@, @deviceSecret@ and @deviceIdentityToken@ attributes are populated, together with any @recipient@-related attributes, to the extent that they exist, from the persisted state.
+** @(RSH8b)@ The @LocalDevice@ @id@ and @deviceSecret@ attributes are generated, and persisted as part of the @LocalDevice@ state, when required by step @(RSH3a2b)@ in the Activation State Machine. At that time, the @clientId@ attribute is also initialised, if the client is identified according to @(RSA7)@.
+** @(RSH8c)@ Following successful registration of a @LocalDevice@, following the procedure in @(RSH3c2a)@, the now known @deviceIdentityToken@ is set and persisted.
+** @(RSH8d)@ If the @LocalDevice* is created by an unidentified client (see @(RSA7)@) and therefore has no @clientId@ set, but the client subsequently becomes identified (as a result of @(RSA7b2)@ or @(RSA7b3)@), then the @LocalDevice@ @clientId@ is set and persisted.
+** @(RSH8e)@ If the @LocalDevice@ @clientId@ becomes set as a result of @(RSH8d)@, and the @LocalDevice@ is already registered (ie the @deviceIdentityToken@ is set), and the ActivationStateMachine is in any state other than @NotActivated@, then a @GotPushDeviceDetails@ event is sent to "the state machine":#RSH3 containing updated @clientId@.
+** @(RSH8f)@ If the @LocalDevice* is created by an unidentified client (see @(RSA7)@) and therefore has no @clientId@ set, but on receipt of a registration response (see @(RSH3c2)@) the registered device has a non-empty @clientId@, then the @LocalDevice@ @clientId@ is set with that @clientId@.
+** @(RSH8g)@ Whenever any change arises of the push transport details for local device (eg an FCM registration token update triggered by the platform), a @GotPushDeviceDetails@ event is sent to "the state machine":#RSH3.
+** @(RSH8h)@ If an attempt to obtain the push transport details for local device (eg an FCM registration token) fails, a @GettingPushDeviceDetailsFailed@ event containing the indicated error is sent to "the state machine":#RSH3.
+** @(RSH8i)@ Each time the library is instanced, if the LocalDevice has push device details (eg an APNS deviceToken), it must verify the validity of those details (eg by requesting a token from the platform and comparing that with the already-known token). If as a result there are updated details, then an update to the Ably server is triggered by sending a @GotPushDeviceDetails@ event to "the state machine":#RSH3.
```

Comment:
I thought this was only needed for APNS?  I can't recall the exact details when we looked at this before, but I thought on Android we could rely on the events being emitted from the OS when the device details changed, but not with APNS.

Reply:
Fixed by af6bbd0c81dc7d15ade879621d79347164cae9ef.